Update Scheduler to have a configurable block provider #7434

seemantaggarwal opened the pull request:
Follow up from https://github.com/paritytech/polkadot-sdk/pull/6362#issuecomment-2629744365

The goal of this PR is to have the scheduler pallet work on a parachain which does not produce blocks on a regular schedule, thus can use the relay chain as a block provider.

Because blocks are not produced regularly, we cannot make the assumption that block number increases monotonically, and thus have new logic to handle multiple spend periods passing between blocks.

Requirement: 

 instead of using the hard coded system block number. We add an associated type BlockNumberProvider

gui1117 commented on substrate/frame/scheduler/src/lib.rs:
```diff
@@ -297,6 +300,8 @@ pub mod pallet {
 	#[pallet::storage]
 	pub type IncompleteSince<T: Config> = StorageValue<_, BlockNumberFor<T>>;
 
+	type BlockNumberProvider: BlockNumberProvider;
```
We will need a doc comment that explains this type, and warns that it must not be too much out of sync with the local block number. More precisely between every block the increment of this number should be reasonably small otherwise performance degrades will degrade a lot as it reads one storage per block number.

seemantaggarwal replied:
Thanks, I will add it towards the end of the pr for sure, I will look into it. At this moment, I am not entirely sure of the value that needs to be advised, i will get more insights into it, thanks for the reminder :)

gui1117 replied:
We can suggest people to use the system block number or the relay chain block number (this second option, only if they expect their parachain to execute very regularly).

gui1117 reviewed (commented):
Direction is good

seemantaggarwal commented:
note: this is a part of https://github.com/paritytech/polkadot-sdk/issues/6297

seemantaggarwal commented:
This is where I am stuck atm: 
`error[E0277]: the trait bound <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber> is not satisfied
     --> /Users/seemantaggarwal/RustroverProjects/polkadot-sdk/substrate/frame/scheduler/src/lib.rs:651:36
      |
  651 | ...                   Lookup::<T>::insert(name, item);
      |                       -------------------       ^^^^ unsatisfied trait bound
      |                       |
      |                       required by a bound introduced by this call`

What I have tried: 
1)  Convert the block number to match expected type: 
something like: `let converted_number: T::BlockNumber = <T::BlockNumberProvider as BlockNumberProvider>::current_block_number().saturated_into();
Lookup::<T>::insert(name, (converted_number, some_value));`
and then something like: 
`if let Some(item) = old::Lookup::<T>::take(id) {
    let converted_item: (u32, u32) = (item.1, item.0.into());  // Convert using `.into()`
    Lookup::<T>::insert(name, converted_item);
}`

2) I have also tried changing the migration functions by adding 
`	pub fn migrate_v2_to_v4() -> Weight  where <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: parity_scale_codec::EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>{
`


Now this, makes me think I have 2 following options going ahead: 

1) Implement a new migrate v4_to_v5 for lookup (I am not sure how to do that yet, but can be figured out)
2) Implement EncodeLike for BlockNumberProvider (as far as I understand this cannot be done since it is coming from codec and not a custom impl)

Would like to get some opinion/help here on the options, or if there is any another thing to try 

Note: things I have tried are just a summary of the rabbit hole,if you do leave a comment here, I can try it out and answer if it works for me or not

gui1117 commented:
> This is where I am stuck atm: 
> ```
> error[E0277]: the trait bound <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber> is not satisfied
>      --> /Users/seemantaggarwal/RustroverProjects/polkadot-sdk/substrate/frame/scheduler/src/lib.rs:651:36
>       |
>   651 | ...                   Lookup::<T>::insert(name, item);
>       |                       -------------------       ^^^^ unsatisfied trait bound
>       |                       |
>       |                       required by a bound introduced by this call
> ```
> 
> What I have tried: 
> 1)  Convert the block number to match expected type: 
> something like: `let converted_number: T::BlockNumber = <T::BlockNumberProvider as BlockNumberProvider>::current_block_number().saturated_into();
> Lookup::<T>::insert(name, (converted_number, some_value));`
> and then something like: 
> `if let Some(item) = old::Lookup::<T>::take(id) {
>     let converted_item: (u32, u32) = (item.1, item.0.into());  // Convert using `.into()`
>     Lookup::<T>::insert(name, converted_item);
> }`
> 
> 2) I have also tried changing the migration functions by adding 
> `	pub fn migrate_v2_to_v4() -> Weight  where <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: parity_scale_codec::EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>{
> `
> 
> 
> Now this, makes me think I have 2 following options going ahead: 
> 
> 1) Implement a new migrate v4_to_v5 for lookup (I am not sure how to do that yet, but can be figured out)
> 2) Implement EncodeLike for BlockNumberProvider (as far as I understand this cannot be done since it is coming from codec and not a custom impl)
> 
> Would like to get some opinion/help here on the options, or if there is any another thing to try 
> 
> Note: things I have tried are just a summary of the rabbit hole,if you do leave a comment here, I can try it out and answer if it works for me or not

The type:
```rust
<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number
```
Doesn't implement:
```rust
EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>
```
Because:
```rust
<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber
```
Is a completely different type.

I think you are trying somewhere to put the system block number into storage, but the type expect the configured provided block number.

seemantaggarwal commented:
Looking at the current failure it fails when running: 
 cargo nextest run --workspace --features runtime-benchmarks benchmark --locked --cargo-profile testnet --cargo-quiet
with the error: 

`Error[E0308]: mismatched types
    --> substrate/frame/scheduler/src/benchmarking.rs:62:40
     |
62   |         Pallet::<T>::do_schedule_named(name, t, period, 0, origin.clone(), call)?;
     |         ------------------------------       ^ expected `sp_runtime::traits::BlockNumberProvider::BlockNumber`, found `sp_runtime::traits::Header::Number`
     |         |
     |         arguments to this function are incorrect
     |
     = note: expected enum `frame_support::traits::schedule::DispatchTime<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>`
                found enum `frame_support::traits::schedule::DispatchTime<<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number>`
note: associated function defined here
    --> substrate/frame/scheduler/src/lib.rs:1047:5
     |
1047 |     fn do_schedule_named(
     |        ^^^^^^^^^^^^^^^^^
1048 |         id: TaskName,
104[9](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:10) |         when: DispatchTime<BlockNumberFor<T>>,
     |         -------------------------------------

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:64:27
    |
64  |     ensure!(Agenda::<T>::get(when).len() == n as usize, "didn't fill schedule");
    |             ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |             |
    |             required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
help: consider further restricting the associated type
    |
55  | ) -> Result<(), &'static str> where <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber> {
    |                               +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0308]: arguments to this function are incorrect
    --> substrate/frame/scheduler/src/benchmarking.rs:164:4
     |
159  |         fill_schedule::<T>(now, s)?;
     |                            ---
     |                            |
     |                            here the type of `now` is inferred to be `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
     |                            here the type of `now` is inferred to be `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
...
164  |             Pallet::<T>::service_agenda(&mut WeightMeter::new(), &mut executed, now, now, 0);
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
note: expected `sp_runtime::traits::BlockNumberProvider::BlockNumber`, found `sp_runtime::traits::Header::Number`
    --> substrate/frame/scheduler/src/benchmarking.rs:164:72
     |
164  |             Pallet::<T>::service_agenda(&mut WeightMeter::new(), &mut executed, now, now, 0);
     |                                                                                 ^^^
     = note: expected associated type `<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber`
                found associated type `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
     = note: an associated type was expected, but a different one was found
note: expected `sp_runtime::traits::BlockNumberProvider::BlockNumber`, found `sp_runtime::traits::Header::Number`
    --> substrate/frame/scheduler/src/benchmarking.rs:164:77
     |
164  |             Pallet::<T>::service_agenda(&mut WeightMeter::new(), &mut executed, now, now, 0);
     |                                                                                      ^^^
     = note: expected associated type `<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber`
                found associated type `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
     = note: an associated type was expected, but a different one was found
note: associated function defined here
    --> substrate/frame/scheduler/src/lib.rs:1186:5
     |
1186 |     fn service_agenda(
     |        ^^^^^^^^^^^^^^
1187 |         weight: &mut WeightMeter,
     |         ------------------------
1188 |         executed: &mut u32,
     |         ------------------
1189 |         now: BlockNumberFor<T>,
     |         ----------------------
1190 |         when: BlockNumberFor<T>,
     |         -----------------------
1191 |         max: u32,
     |         --------

error[E0308]: mismatched types
   --> substrate/frame/scheduler/src/benchmarking.rs:298:22
    |
135 | #[benchmarks]
    | ------------- arguments to this function are incorrect
...
295 |         fill_schedule::<T>(when, s)?;
    |                            ---- here the type of `when` is inferred to be `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
...
298 |         _(RawOrigin::Root, when, periodic, priority, call);
    |                            ^^^^ expected `sp_runtime::traits::BlockNumberProvider::BlockNumber`, found `sp_runtime::traits::Header::Number`
    |
    = note: expected associated type `<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber`
               found associated type `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    = note: an associated type was expected, but a different one was found
note: associated function defined here
   --> substrate/frame/scheduler/src/lib.rs:396:[10](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:11)
    |
396 |         pub fn schedule(
    |                ^^^^^^^^
397 |             origin: OriginFor<T>,
398 |             when: BlockNumberFor<T>,
    |             -----------------------

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:300:28
    |
300 |         ensure!(Agenda::<T>::get(when).len() == s as usize + 1, "didn't add to schedule");
    |                 ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                 |
    |                 required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
help: consider further restricting the associated type
    |
288 |     ) -> Result<(), BenchmarkError> where <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber> {
    |                                     +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:300:28
    |
300 |         ensure!(Agenda::<T>::get(when).len() == s as usize + 1, "didn't add to schedule");
    |                 ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                 |
    |                 required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:310:31
    |
310 |         assert_eq!(Agenda::<T>::get(when).len(), s as usize);
    |                    ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                    |
    |                    required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
help: consider further restricting the associated type
    |
306 |     fn cancel(s: Linear<1, { T::MaxScheduledPerBlock::get() }>) -> Result<(), BenchmarkError> where <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber> {
    |                                                                                               +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0308]: mismatched types
   --> substrate/frame/scheduler/src/benchmarking.rs:315:41
    |
[13](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:14)5 | #[benchmarks]
    | ------------- arguments to this function are incorrect
...
309 |         fill_schedule::<T>(when, s)?;
    |                            ---- here the type of `when` is inferred to be `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
...
3[15](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:16) |         _(schedule_origin as SystemOrigin<T>, when, 0);
    |                                               ^^^^ expected `sp_runtime::traits::BlockNumberProvider::BlockNumber`, found `sp_runtime::traits::Header::Number`
    |
    = note: expected associated type `<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber`
               found associated type `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    = note: an associated type was expected, but a different one was found
note: associated function defined here
   --> substrate/frame/scheduler/src/lib.rs:418:10
    |
418 |         pub fn cancel(origin: OriginFor<T>, when: BlockNumberFor<T>, index: u32) -> DispatchResult {
    |                ^^^^^^                       -----------------------

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:323:31
    |
323 |             s == 1 || Agenda::<T>::get(when)[0].is_none(),
    |                       ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                       |
    |                       required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:[17](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:18)1:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
help: consider further restricting the associated type
    |
306 |     fn cancel(s: Linear<1, { T::MaxScheduledPerBlock::get() }>) -> Result<(), BenchmarkError> where <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber> {
    |                                                                                               +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:327:30
    |
327 |             s > 1 || Agenda::<T>::get(when).len() == 0,
    |                      ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                      |
    |                      required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:[21](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:22)
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
help: consider further restricting the associated type
    |
306 |     fn cancel(s: Linear<1, { T::MaxScheduledPerBlock::get() }>) -> Result<(), BenchmarkError> where <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber> {
    |                                                                                               +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:310:31
    |
310 |         assert_eq!(Agenda::<T>::get(when).len(), s as usize);
    |                    ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                    |
    |                    required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:3[23](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:24):31
    |
323 |             s == 1 || Agenda::<T>::get(when)[0].is_none(),
    |                       ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                       |
    |                       required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:327:30
    |
327 |             s > 1 || Agenda::<T>::get(when).len() == 0,
    |                      ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                      |
    |                      required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`

error[E0308]: mismatched types
   --> substrate/frame/scheduler/src/benchmarking.rs:348:[26](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:27)
    |
135 | #[benchmarks]
    | ------------- arguments to this function are incorrect
...
345 |         fill_schedule::<T>(when, s)?;
    |                            ---- here the type of `when` is inferred to be `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
...
348 |         _(RawOrigin::Root, id, when, periodic, priority, call);
    |                                ^^^^ expected `sp_runtime::traits::BlockNumberProvider::BlockNumber`, found `sp_runtime::traits::Header::Number`
    |
    = note: expected associated type `<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber`
               found associated type `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    = note: an associated type was expected, but a different one was found
note: associated function defined here
   --> substrate/frame/scheduler/src/lib.rs:428:10
    |
428 |         pub fn schedule_named(
    |                ^^^^^^^^^^^^^^
...
431 |             when: BlockNumberFor<T>,
    |             -----------------------

error[E0[27](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:28)7]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:350:[28](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:29)
    |
350 |         ensure!(Agenda::<T>::get(when).len() == s as usize + 1, "didn't add to schedule");
    |                 ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                 |
    |                 required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
help: consider further restricting the associated type
    |
337 |     ) -> Result<(), BenchmarkError> where <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber> {
    |                                     +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:350:28
    |
350 |         ensure!(Agenda::<T>::get(when).len() == s as usize + 1, "didn't add to schedule");
    |                 ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                 |
    |                 required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:372:31
    |
372 |             s == 1 || Agenda::<T>::get(when)[0].is_none(),
    |                       ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                       |
    |                       required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
help: consider further restricting the associated type
    |
358 |     ) -> Result<(), BenchmarkError> where <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber> {
    |                                     +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:376:[30](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:31)
    |
376 |             s > 1 || Agenda::<T>::get(when).len() == 0,
    |                      ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                      |
    |                      required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
help: consider further restricting the associated type
    |
358 |     ) -> Result<(), BenchmarkError> where <<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber> {
    |                                     +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:372:[31](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:32)
    |
372 |             s == 1 || Agenda::<T>::get(when)[0].is_none(),
    |                       ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                       |
    |                       required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`

error[E0277]: the trait bound `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number: EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not satisfied
   --> substrate/frame/scheduler/src/benchmarking.rs:[37](https://github.com/paritytech/polkadot-sdk/actions/runs/13163415501/job/36737726646#step:4:38)6:30
    |
376 |             s > 1 || Agenda::<T>::get(when).len() == 0,
    |                      ---------------- ^^^^ the trait `EncodeLike<<<T as pallet::Config>::BlockNumberProvider as sp_runtime::traits::BlockNumberProvider>::BlockNumber>` is not implemented for `<<<T as frame_system::Config>::Block as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number`
    |                      |
    |                      required by a bound introduced by this call
    |
note: required by a bound in `frame_support::pallet_prelude::StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get`
   --> /__w/polkadot-sdk/polkadot-sdk/substrate/frame/support/src/storage/types/map.rs:171:21
    |
171 |     pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> QueryKind::Query {
    |                        ^^^^^^^^^^^^^^^ required by this bound in `StorageMap::<Prefix, Hasher, Key, Value, QueryKind, OnEmpty, MaxValues>::get``

seemantaggarwal commented:
Reverted the last commit, as it did not really solve much, bringing it to attention of @gui1117  @ggwpez 
i fixed the test too though. 

https://github.com/paritytech/polkadot-sdk/pull/7441#issuecomment-2639029169 -> read the comment here

muharem commented:
> Error[E0308]: mismatched types --> substrate/frame/scheduler/src/benchmarking.rs:62:40 | 62   |         Pallet::<T>::do_schedule_named(name, t, period, 0, origin.clone(), call)?; |         ------------------------------       ^ expected sp_runtime::traits::BlockNumberProvider::BlockNumber, found sp_runtime::traits::Header::Number|         |

@seemantaggarwal should be your new block number alias instead 

https://github.com/paritytech/polkadot-sdk/blob/f373af0d1c1e296c1b07486dd74710b40089250e/substrate/frame/scheduler/src/benchmarking.rs#L53

seemantaggarwal commented:
Requesting review from @gui1117  @muharem @ggwpez @kianenigma

ggwpez commented on substrate/frame/scheduler/src/lib.rs:
```diff
@@ -292,6 +294,8 @@ pub mod pallet {
 		/// The preimage provider with which we look up call hashes to get the call.
 		type Preimages: QueryPreimage<H = Self::Hashing> + StorePreimage;
+
+		type BlockNumberProvider: BlockNumberProvider;
```
Docs? Devs using this pallet wont know what it is or how it should be used.

seemantaggarwal replied:
addressed in commit 03fc0dad9059032ad884c1da2024a58260b6571d

ggwpez replied:
Please check how we normally write these docs. This is a comment `//` instead of a rust doc.  
Also there should be one introduction sentence on what it does and then more elaborate paragraph following.  

Just put yourself into the position of someone having never seen this before. What is it? How am i supposed to use it? And possible implications.

seemantaggarwal replied:
addressed in commit -> d7a8db1832c06e330abc6afb29a87ff3131325e0
open question, I did try and put myself into the shoes for somebody new (was relatively easy, did not require any roleplaying haha) and I think this should be maybe a part of a larger readme? Not just for this pallet, but over multiple pallets explaining this?

gui1117 replied:
Runtime will implement the config trait they usually want documentation as precise as possible.
I agree crate documentation is also good, and in this PR case we can warn in both the crate doc and the config doc.

ggwpez replied:
added some docs now

kianenigma replied:
@seemantaggarwal adding documentation to public items in Rust is a must, please follow the advise given by other peers in this regard.

ggwpez commented on prdoc/pr_7441.prdoc:
```diff
@@ -0,0 +1,26 @@
+
+    The goal of this PR is to have the scheduler pallet work on a parachain which does not produce blocks on a regular schedule, thus can use the relay chain as a block provider.
+
+    Because blocks are not produced regularly, we cannot make the assumption that block number increases monotonically, and thus have new logic to handle multiple spend periods passing between blocks.
```
This is out dated now as we only add the new config item.

ggwpez commented on prdoc/pr_7441.prdoc:
```diff
@@ -0,0 +1,26 @@
+- name: pallet-referenda
+  bump: minor
+- name: pallet-scheduler
+  bump: minor
```
Adding a config Item is a breaking change, thought the CI would catch that.

seemantaggarwal replied:
addressed in commit https://github.com/paritytech/polkadot-sdk/commit/03fc0dad9059032ad884c1da2024a58260b6571d

seemantaggarwal commented on substrate/frame/scheduler/src/tests.rs:
```diff
@@ -1636,6 +1636,7 @@ fn on_initialize_weight_is_correct() {
 		// Will include the named periodic only
+		System::set_block_number(1);
 		assert_eq!(
 			Scheduler::on_initialize(1),
```
I see, thanks, I need to start it by 0 instead of 1, and just add a comment, that this should not make a difference to the working? I did not understand what is the comment I need to put

ggwpez commented on substrate/frame/scheduler/src/lib.rs:
```diff
@@ -374,9 +382,10 @@ pub mod pallet {
+	impl<T: Config> Hooks<SystemBlockNumberFor<T>> for Pallet<T> {
 		/// Execute the scheduled calls
-		fn on_initialize(now: BlockNumberFor<T>) -> Weight {
+		fn on_initialize(_do_not_use_local_block_number: SystemBlockNumberFor<T>) -> Weight {
```
In this case we want to always use the injected block number provider, so i think it makes sense.

ggwpez commented on substrate/frame/scheduler/src/migration.rs:
```diff
@@ -19,7 +19,6 @@
-use frame_system::pallet_prelude::BlockNumberFor;
 
 #[cfg(feature = "try-runtime")]
 use sp_runtime::TryRuntimeError;
```
We should probably deprecate the migrations. They would not interact so well with changing the BN provider at the same time and it is a lot of legacy code

ggwpez reviewed (approved):
LGTM, had to push some fixes since we want to get this into the next release.
